### PR TITLE
Enables distributed GCR

### PR DESCRIPTION
**Pull request by MarcelKoch**
This PR enables the GCR for distributed vectors.

### CODE REVIEW

File: core/distributed/vector.cpp [hunk 1]
```diff
@@ -517,28 +537,28 @@ void Vector<ValueType>::compute_norm2(ptr_param<LinOp> result,
...
 }
 
 
 template <typename ValueType>
-void Vector<ValueType>::compute_norm1(ptr_param<LinOp> result) const
+void Vector<ValueType>::compute_squared_norm2(ptr_param<LinOp> result) const
 {
     array<char> tmp{this->get_executor()};
-    this->compute_norm1(result, tmp);
+    this->compute_norm2(result, tmp);
```
Comment (yhmtsai) [inline core/distributed/vector.cpp:547]: compute_sqared_norm2?

File: core/distributed/vector.cpp [hunk 2]
```diff
@@ -498,14 +498,34 @@ void Vector<ValueType>::compute_norm2(ptr_param<LinOp> result) const
 template <typename ValueType>
 void Vector<ValueType>::compute_norm2(ptr_param<LinOp> result,
                                       array<char>& tmp) const
+{
+    using NormVector = typename local_vector_type::absolute_type;
+    auto exec = this->get_executor();
+    const auto comm = this->get_communicator();
+    auto dense_res = make_temporary_clone(exec, as<NormVector>(result));
+    compute_squared_norm2(dense_res.get(), tmp);
```
Comment (yhmtsai) [inline core/distributed/vector.cpp:506]:
```suggestion
    this->compute_squared_norm2(dense_res.get(), tmp);
```
nit

**Review by yhmtsai — CHANGES_REQUESTED** [review]

File: core/distributed/helpers.hpp [hunk 3]
```diff
@@ -168,6 +168,50 @@ void vector_dispatch(T* linop, F&& f, Args&&... args)
 }
 
 
+/**
+ * Helper to extract a submatrix.
+ *
+ * @note  global_size is unused, since it can be deferred from rows and cols.
```
Comment (tcojean) [inline core/distributed/helpers.hpp:174]: nit
```suggestion
 * @note  global_size is unused, since it can be inferred from rows and cols.
```

**Review by tcojean — APPROVED** [review]
LGTM

**Review by upsj — APPROVED** [review]
LGTM!

**Review by yanjen — APPROVED** [review]
LGTM, but I have a question and a suggestion.
Thank you for implementing the distributed version. I was too caught up in ICON and did not look into this.

File: core/solver/gcr.cpp [hunk 4]
```diff
@@ -122,19 +123,23 @@ void Gcr<ValueType>::apply_dense_impl(const VectorType* dense_b,
 
     const auto num_rows = this->get_size()[0];
     const auto num_rhs = dense_b->get_size()[1];
+    const auto local_num_rows =
```
[the remark this answers is not in the document]
Comment (MarcelKoch) [reply core/solver/gcr.cpp:126]: In the distributed setting, the matrix becomes too large for a single process, so we have to distribute the matrix. In some cases, where you have relatively small matrices and many many right-hand-sides, your suggestion makes sense. But that is not what we consider right now.

**Pull request merged**